sycl/doc/OptionalDeviceFeatures.md

```diff
@@ -0,0 +1,698 @@
+starts with a string giving the name of the type which is followed by a list of
+`i32` constants where each constant is a value from `enum class aspect` telling
+the numerical value of an aspect from the type's
+`[[sycl_detail::uses_aspects()]]` attribute.  In the example above, the type
+`cl::sycl::detail::half_impl::half` uses an aspect whose numerical value is
+`8` and the type `cl::sycl::other_type` uses two aspects `8` and `9`.
+
+**NOTE**: The reason we choose this representation is because LLVM IR does not
+allow metadata to be attached directly to types.  This representation works
+around that limitation by creating global named metadata that references the
+type's name.
+
+We also introduce two metadata that can be attached to a function definition
+similar to the existing `!intel_reqd_sub_group_size`.  The
+`!intel_declared_aspects` metadata is used for functions that are decorated
+with `[[sycl::requires()]]`, and the `!intel_used_aspects` metadata is used to
+store the propagated information about all aspects used by a kernel or exported
+device function.
+
+In each case, the metadata's parameter is an unnamed metadata node, and the
+value of the metadata node is a list of `i32` constants, where each constant is
+a value from `enum class aspect`.
+
+For example, the following illustrates the IR that corresponds to a function
+`foo` that is decorated with `[[sycl::requires()]]` where the required aspects
+have the numerical values `8` and `9`.  In addition, the function uses an
+optional feature that corresponds to an aspect with numerical value `8`.
+
+```
+define void @foo() !intel_declared_aspects !1 !intel_used_aspects !2 {}
+!1 = !{i32 8, i32 9}
+!2 = !{i32 8}
+```
+
+
+### Changes to the DPC++ front-end
+
+The front-end of the device compiler is responsible for parsing the
+`[[sycl::requires()]]` and `[[sycl_detail::uses_aspects()]]` attributes and
+transferring the information to the LLVM IR metadata described above according
+to the following rules:
+
+* If the translation unit contains any type definitions that are decorated with
+  `[[sycl_detail::uses_aspects()]]`, the front-end creates an
+  `!intel_types_that_use_aspects` metadata describing the aspects used by all
+  such types.
+
+* If a function is decorated with `[[sycl_detail::uses_aspects()]]`, the
+  front-end adds an `!intel_used_aspects` metadata to the function's definition
+  listing the aspects from that attribute.
+
+* If a function is decorated with `[[sycl::requires()]]`, the front-end adds
+  an `!intel_declared_aspects` metadata to the function's definition listing
+  the aspects from that attribute.
+
+
+### New LLVM IR pass to propagate aspect usage
+
+We add a new IR phase to the device compiler which does the following:
+
+* Creates (or augments) a function's `!intel_used_aspects` metadata with
+  aspects that come from references to types in the
+  `intel_types_that_use_aspects` list.
```

Comment:
@AaronBallman, @erichkeane, how hard/easy/feasible would it be to populate list of used aspects with ones which are attached to types used within a function in FE?

I'm asking, because doing so at LLVM IR level would be expensive, I think. My concerns are:
- AFAIK there are no way to easily grab a list of all types used within a function, which means that we will have to go through every argument of every instruction within a function and check its type
- Since types are encoded in metadata as strings, we will have to perform not-so-cheap string comparison to understand whether it is a type we are looking for or not
- According to [`llvm::Type`](https://llvm.org/doxygen/classllvm_1_1Type.html) class definition, only struct types have names, which means that if we need to attach an aspect to a non-aggregate type or to an aggregate other than struct, then we have to come up with some naming scheme for that. @gmlueck, I assume that we will anyway provide access to some "native" types through some wrappers/aliases, so we should be always able to annotate that wrapper. However, some optimizations could unwrap the type and pass it directly - in that case we will loose the annotation

Reply:
I agree that encoding the type reference in the metadata as a string seems kind of gross.  I think @erichkeane said that there is some precedent for doing this in clang, though.

How do LLVM IR instructions reference a type?  I assume this is done through some sort of ID?  I presume IR instructions do not reference a type via a string.  Is it possible for a value in a metadata node to be one of these type IDs?  It seems like that would be a lot more efficient than referencing by string.

Reply:
@AlexeySachkov : Detecting all uses within a function is quite difficult, we'd have to go through every single level of the AST with some custom code to detect when/where that is used.  It becomes REALLY difficult/time consuming to do in the CFE.

Reply:
> How do LLVM IR instructions reference a type? I assume this is done through some sort of ID? I presume IR instructions do not reference a type via a string. Is it possible for a value in a metadata node to be one of these type IDs? It seems like that would be a lot more efficient than referencing by string.

They are referenced through pointer to a corresponding class. I guess we are able to optimize the search by retrieving a `Type *` based on a string encoded into a metadata and later comparing it with `Type *` encoded in each instruction; but that would still require analysis of each instruction of each function in a module

Reply:
That's a good suggestion to optimize the search.  I added it as a note, so we don't forget it.  See 29a2e9f504372d40c1487a569f95b4079d1ab4ba.